### PR TITLE
Feat/granite guardian prompt injection

### PR DESCRIPTION
**Pull request by ovindumandith**
## Purpose
Adds a new AI guardrail policy to detect and block prompt injection and jailbreak attempts in LLM API requests before they reach the upstream model. Without this policy, adversarial users can craft inputs that override system instructions, bypass safety constraints, or extract sensitive context from the LLM's system prompt.

## Goals
* **Intercept & Classify:** Provide a reusable gateway policy that intercepts LLM requests and classifies them using IBM Granite Guardian 3.3 8B.
* **Gateway-Level Blocking:** Block flagged requests at the gateway level - no changes needed to the upstream LLM service.
* **Operator Control:** Give operators control over sensitivity (confidence threshold), failure behavior (fail-open/fail-closed), and response format.

## Approach
* The policy is implemented as a Python gateway policy following the same structure as the existing `prompt-compressor` policy (`apip_sdk_core` SDK, `src/` layout, stub-based unit tests, `pyproject.toml` packaging).
* **On each request:**
  * The request body is buffered and the user message is extracted via a configurable JSONPath expression.
  * For each configured risk category, an OpenAI-compatible chat completion request is sent to a Granite Guardian inference endpoint with the risk name embedded in the system message.
  * The model returns a `<score>Yes/No</score>` verdict; log-probability is used to derive a confidence score.
  * If any risk is detected above the configured threshold, the request is rejected with a structured JSON error; otherwise, it passes through unchanged.
* **Files added:**
  * `policies/granite_guardian_prompt_injection/` - Full policy package (`policy.py`, `pyproject.toml`, `policy-definition.yaml`, `requirements.txt`, `tests/`).
  * `docs/granite-guardian-prompt-injection/v0.1.0/`   Policy documentation and metadata.

## User Stories
* **As an API platform operator**, I want to block prompt injection attempts at the gateway so that my upstream LLM cannot be manipulated by adversarial inputs.
* **As an API platform operator**, I want to configure which risk categories to check and a confidence threshold so that I can tune the trade-off between security and false positives.
* **As an API platform operator**, I want the guardrail to fail closed by default so that my LLM is protected even if the guardrail service is temporarily unavailable.

## Release Note
**Added `granite-guardian-prompt-injection` v0.1.0** - A new Python policy that detects prompt injection and jailbreak attempts in LLM API requests using IBM Granite Guardian 3.3 8B. Requests flagged above the configured confidence threshold are rejected before reaching the upstream LLM. Supports configurable risk categories, confidence thresholds, block status codes, and fail-open/fail-closed error handling.

## Documentation, Training, & Marketing
* **Documentation:** Granite Guardian Prompt Injection - v0.1.0 (included in this PR)
* **Training:** N/A
* **Certification:** N/A - new infrastructure policy, no impact on existing certification exams
* **Marketing:** N/A

### Automation Tests
* **Unit tests:** 41 unit tests covering processing mode, system/request param normalisation, JSONPath resolution, passthrough edge cases, blocking logic, confidence threshold gating, assessment response format, multi-risk evaluation order, error handling (fail-open/closed), and HTTP call details.
  * *Run:* `python -m unittest discover -s tests -p 'test_*.py' -v` inside the `policies/granite_guardian_prompt_injection/` directory.
  * *Note:* All external dependencies (`apip_sdk_core`, `requests`) are stubbed - no network calls in tests.
* **Integration tests:** N/A for this PR - policy is self-contained and has no integration test coverage yet.

### Test Environment
* Python 3.10+
* Windows 11 (local development)
* No external services required for unit tests (all dependencies stubbed)

## Security Checks
- [x] Followed secure coding standards in WSO2 Secure Engineering Guidelines? **Yes**
- [ ] Ran FindSecurityBugs plugin and verified report? **N/A (Python policy)**
- [x] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? **Yes**

## Samples
See **Reference Scenarios** in the documentation:
1. Basic setup protecting a chat completions route with default risks (`jailbreak`, `prompt_injection`).
2. Strict detection with lower threshold and `showAssessment: true` for detailed block responses.
3. Fail-open configuration for high-availability deployments.

## Related PRs & Migrations
* **Related PRs:** N/A
* **Migrations:** N/A — new policy, no migration required.

## Learning & References
* [ibm-granite/granite-guardian-3.3-8b · Hugging Face](https://huggingface.co/ibm-granite/granite-guardian-3.3-8b) — Risk category definitions and `<guardianconfig>` system message format.
* Log-probability confidence extraction pattern from vLLM's `logprobs` API response.

### CODE REVIEW

File: policies/granite-guardian-prompt-injection/src/granite_guardian_prompt_injection_v0/__init__.py
Comment (renuka-fernando) [inline policies/granite-guardian-prompt-injection/src/granite_guardian_prompt_injection_v0/__init__.py:1]: PR check is failing.

```
1.863 ERROR: [Discovery] policy name mismatch: build file declares 'granite_guardian_prompt_injection' but policy-definition.yaml has 'granite-guardian-prompt-injection' at /tmp/pip-policy-896557785/granite_guardian_prompt_injection_v0
```

Let's rename the directory.

```
policies/granite_guardian_prompt_injection/ -> policies/granite-guardian-prompt-injection/

Comment (renuka-fernando) [reply policies/granite-guardian-prompt-injection/src/granite_guardian_prompt_injection_v0/__init__.py:1]: Do the same for docs dir.

File: policies/granite_guardian_prompt_injection/src/granite_guardian_prompt_injection_v0/policy.py [hunk 1]
```diff
@@ -0,0 +1,359 @@
+# Copyright (c) 2025, WSO2 LLC. (https://www.wso2.com).
```
Comment (renuka-fernando) [inline policies/granite_guardian_prompt_injection/src/granite_guardian_prompt_injection_v0/policy.py]:
```suggestion
# Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
```

File: policies/granite_guardian_prompt_injection/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+[tool.hatch.build.targets.wheel.force-include]
+"policy-definition.yaml" = "granite_guardian_prompt_injection_v0/policy-definition.yaml"
+
+[tool.hatch.build.targets.sdist]
+only-include = [
+  "src/granite_guardian_prompt_injection_v0",
+  "pyproject.toml",
+  "policy-definition.yaml",
+  "README.md",
+]
```
Comment (renuka-fernando) [inline policies/granite_guardian_prompt_injection/pyproject.toml]: plz add a new line at the end of the file.

File: policies/granite-guardian-prompt-injection/src/granite_guardian_prompt_injection_v0/policy.py [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+            "logprobs": True,
+            "top_logprobs": 5,
+        }
+
+        response = http_client.post(
+            f"{self._params.endpoint}/v1/chat/completions",
+            headers=headers,
+            json=payload,
+            timeout=self._params.timeout,
+        )
```
Comment (renuka-fernando) [inline policies/granite-guardian-prompt-injection/src/granite_guardian_prompt_injection_v0/policy.py:152]: If this policy just makes an HTTP request, we can write this policy as a Go policy, right?
cc: @Arshardh

**Review by renuka-fernando — COMMENTED** [review]
Hi @ovindumandith 

Thanks for the contribution. I have some comments. Could you please check?

File: policies/granite-guardian-prompt-injection/src/granite_guardian_prompt_injection_v0/__init__.py
Comment (ovindumandith) [reply policies/granite-guardian-prompt-injection/src/granite_guardian_prompt_injection_v0/__init__.py:1]: Fixed in latest commit, thakyou

File: policies/granite_guardian_prompt_injection/src/granite_guardian_prompt_injection_v0/policy.py (hunk 1 above)
Comment (ovindumandith) [reply policies/granite_guardian_prompt_injection/src/granite_guardian_prompt_injection_v0/policy.py]: Fixed in latest commit

File: policies/granite_guardian_prompt_injection/pyproject.toml (hunk 2 above)
Comment (ovindumandith) [reply policies/granite_guardian_prompt_injection/pyproject.toml]: Fixed in latest commit

File: policies/granite-guardian-prompt-injection/src/granite_guardian_prompt_injection_v0/policy.py (hunk 3 above)
Comment (ovindumandith) [reply policies/granite-guardian-prompt-injection/src/granite_guardian_prompt_injection_v0/policy.py:152]: Yes, technically this policy could be written in Go since the model inference runs on an external OpenAI compatible endpoint and the policy itself only makes an HTTP call. We initially went with Python following the same pattern as prompt-compressor, which also handles model interactions though in that case Python is genuinely needed for the compression-prompt NLP library.

**Review by renuka-fernando — APPROVED** [review]
